### ISSUE TITLE
Link Express and Socket.IO

### ISSUE THREAD

**Issue by shimaore** [zappajs/zappajs#32]
[Follow-up from conversation in #8 ]

@scien the issue as I understand it is that the (server-side) sessions inside Express and inside Socket.IO are independent, so for example you can't do:

```
@get '/emit': ->
    # Express session has access to the proper socket for that client.
    @emit 'username':@session.username

@on 'connect': ->
   # Socket.IO has access to the Express session.
   @emit 'username': @session.username
```

So the idea (at least) would to be able to provide `@emit` inside an Express context, and `@session` inside a Socket.IO context.

The [solution I referenced](http://stackoverflow.com/questions/4753957/socket-io-authentication/4782933#4782933) essentially does:
1. client first connects to the socket server (to obtain socket ID);
2. client queries Zappa (via Express) (with socket ID and desired Socket.IO channel name); Zappa returns authentication token and stores the mapping between the socket ID and the Express session (so that it can retrieve the Express session from the socket session, server-side); and stores the session ID in the Express session.
3. client subscribes to channel using the authentication token.

**Comment by scien** [comment]
http://www.danielbaulig.de/socket-ioexpress/

This article made it seem easier than that.  It looked like we could just create a MemoryStore and store a reference to it in the session middleware which would then be accessible in sockets, so that could be synced up upon handshake.

```
var io = require('socket.io'),
    express = require('express'),
    MemoryStore = express.session.MemoryStore,
    app = express.createServer(),
    sessionStore = new MemoryStore();

app.configure(function () {
    app.use(express.cookieParser());
    app.use(express.session({store: sessionStore
        , secret: 'secret'
        , key: 'express.sid'}));
    app.use(function (req, res) {
        res.end('<h2>Hello, your session id is ' + req.sessionID + '</h2>');
    });
});
```

and then store a map of connections somewhere as well that could be referenced by express

```
var sessionsConnections = {};
sio.on('connection', function (socket) {
    // do all the session stuff
    sessionsConnections[socket.handshake.sessionID] = socket;
});

app.get('/', function (req, res) {
    sio.sockets.in(req.sessionID).send('Man, good to see you back!');
});
```

This solution would be able to have everything setup on the handshake instead of requiring the extra request and subscribes steps.

As I understand (correct me if I'm missing something), the your referencing looks like it just provides authorization for web sockets (which is still a useful feature).

The solution I referenced (at the top here) seems like it would only work if we're running express + socket.io on the same server (access to the MemoryStore).  In many cases (mine included), someone might run these on separate servers.  in this case, it seems like we just need 1. get the socket id to express server, 2. get the session id to io server 3. and have some session storage available to both (most likely redis for pub/sub from express to socket.io - http://howtonode.org/redis-pubsub).

Is there any reason we shouldn't approach these problems separately? 1. express being able to emit to a socket. 2. socket.io having access to session data.
1. could be accomplished by telling express the socketid after connection, then having express open its own connection to io to publish on or pushing events through an api (http://pusher.com/docs/server_api_guide/server_publishing_events) like pusherapp.
2. could be accomplished by telling socket.io the session id, then using any pluggable shared storage (redis, mongo, mysql, s3, etc.)

Thoughts?

**Comment by shimaore** [comment]
The solution on danielbaulig.de is all in memory so I'd guess that (a) it will badly leak and won't scale, (b) can't be implemented with separate view and sockets servers. More importantly (if I read properly the excerpts you mentioned) they provide the client with the Express session ID, which is bad security wise.

I've started working on this in branch `issue_32` (commit d1b16eefa73d9e2d4e0eb2d9862904c11a1c29d2); I've tried to use the native stores for Express/Socket.IO to avoid the need to add another store -- but if that's what we need to get it to work then that's what we'll do. (They both support `redis` though, so that shouldn't be an issue.)

I've implemented the authorization `key`, for now I left the authorization piece on channel subscription out.

I haven't tested that code at all, feel free to give it a run and tear it apart if needed; I'll pick this up tomorrow.

**Comment by scien** [comment]
Here's the approach I was thinking (part 1) - https://github.com/scien/express-io-pubsub.  Two servers connect to either Redis or MongoDB, then the express side can publish to a room, and socket.io side can subscribe to receive the messages in near real-time. right now it just broadcasts to a room, but if we had the authorization piece you started to work on, express could send the socket id to emit to instead of a room name.

then the other way around (part 2), express could become the "publisher" and store session data in redis or mongo capped collection. then auth, then socket.io can read by session id.

**Comment by shimaore** [comment]
I think my goal is simpler, but would allow to implement the protocol you described, or any number of other protocols.

Basically I'm trying to achieve the following, using the existing pluggable stores of both Express and Socket.IO:
- if Express and Socket.IO use (at least) a MemoryStore (the default for both), then if you use a single Zappa server you can (a) access the socket (e.g. write to the client asynchronously using Socket.IO) from Express, and (b) you can access the Express session data from Socket.IO; obviously this would still work if either or both Express and Socket.IO are using an external store (Redis, ...);
- if Express and Socket.IO use external stores (e.g. Redis), then if a client connects to a Zappa instance, you can access the Express session data (where Express is running in a different process/server) from the client's Socket.IO connection; you may also be able to have Express send data to the existing Socket.IO connection, although I'm unclear what that means (=Express can retrieve the socket ID, but I don't know WebSockets well enough to know whether the client will recognize the messages coming from Express as the same as its exisiting connection to Socket.IO on a different server).
- both Express and Socket.IO may challenge, independently, the client for some authentication token;
- the session ID is not accessible to scripts on the client (since that would open security risks).

How Express and Socket.IO on different servers communicate together (in the back-end) is left to the developer -- I don't think that belongs in Zappa.

Obviously since I don't have a clear idea of what a use case may be I might be missing some pieces of that puzzle.

**Comment by shimaore** [comment]
Use case from original request:

> Let regular requests that hit Express actions `emit` messages back to clients
> over all socket.io connections they've established (with the same Express session id).
> This is great for handling some work asynchronously in Node (like file parsing) by
> sending back a response immediately, then sending updates to the browser as events
> occur during your long task (like 'finished parsing', or 'error parsing').

**Comment by scien** [comment]
Cool. That all sounds good.  on your second point, i don't think you'd ever send the websocket message from express (there wouldn't be any connection open, unless you wanted socket.io running on both servers which kind of defeats the point. you just want express to be able to tell socket.io to send a specific message).  I can at least share a few use cases that came to mind for my purposes.
- When a user plays a song on my site, I send an http request to a server to log the action.  This action could then be passed on to socket.io to broadcast the event to all the users friends (saving the process of sending a response back to the user, and sending out a separate websocket request)
- Outside of express... I could have a script running to look at who's online and what they like, come up with some recommendations of what they should do, then I could publish events to socket.io to get broadcast to users online to help keep them more engaged.

my middleware solution above will let me handle both these cases for now.  let me know if other uses cross your mind.

**Comment by shimaore** [comment]
I just merged the `issue_32` branch into `master`.
Please check out `examples/share_express.coffee` and `examples/share_socket.coffee`.

**Comment by shimaore** [comment]
Fixed in 0.4.7
